src/approaches/grammar_search_invention_approach.py

```diff
@@ -702,6 +705,82 @@ def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
             score += -trans_log_prob  # remember that lower is better
         return score
 
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _HAddHeuristicBasedScoreFunction(_HeuristicBasedScoreFunction):  # pylint:disable=abstract-method
+    """Implement _generate_heuristic() with HAdd.
+    """
+    def _generate_heuristic(self, init_atoms: Set[GroundAtom],
+                            objects: Set[Object],
+                            goal: Set[GroundAtom],
+                            strips_ops: Sequence[STRIPSOperator],
+                            option_specs: Sequence[OptionSpec],
+                            ground_ops: Set[_GroundSTRIPSOperator]
+                            ) -> Callable[[Set[GroundAtom]], float]:
+        relaxed_operators = frozenset({utils.RelaxedOperator(
+            op.name, utils.atoms_to_tuples(op.preconditions),
+            utils.atoms_to_tuples(op.add_effects)) for op in ground_ops})
+        hadd_fn = utils.HAddHeuristic(
+            utils.atoms_to_tuples(init_atoms),
+            utils.atoms_to_tuples(goal),
+            relaxed_operators)
+        def _hadd_fn_h(init_atoms: Set[GroundAtom]) -> float:
+            return hadd_fn(utils.atoms_to_tuples(init_atoms))
+        return _hadd_fn_h
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _ExactHeuristicBasedScoreFunction(_HeuristicBasedScoreFunction):  # pylint:disable=abstract-method
+    """Implement _generate_heuristic() with task planning.
```

Comment:
just curious, what's going on with the pylint thing here? i don't see any abstract methods

Reply:
pylint complains when you make a class that only partially implements an abstract parent class. another solution is to copy-paste the abstract method from parent, and keep it abstract here